File: BOJ/24_DFS와BFS/2667_단지번호붙이기.py
# 얘도 dfs 돌면 되는건데 continue 생각을 잘못했네ㅠ
def dfs(x,y) :
    global cnt,N
    visited[x][y] = True
    for (dx,dy) in zip(dxs,dys) :
        if x+dx < 0 or x+dx >= N or y+dy < 0 or y+dy >= N : continue
        if board[x+dx][y+dy] == '1' and visited[x+dx][y+dy] == False:
            cnt += 1
            dfs(x+dx,y+dy)


N = int(input())
board = [input() for _ in range(N)]
visited = [[False] * N for _ in range(N)]
dxs = [-1,1,0,0]
dys = [0,0,-1,1]
danji = []
for x in range(N) :
    for y in range(N) :
        if board[x][y] =='1' and visited[x][y] == False :
            cnt = 1
            dfs(x,y)
            danji.append(cnt)

danji.sort()
print(len(danji))
for d in danji :
    print(d)

File: BOJ/24_DFS와BFS/2667_단지번호붙이기.cpp
#include <cstdio>
#include <iostream>
#include <vector>
#include <string.h>
#include <algorithm>
using namespace std;

int N,cnt;
bool visited[26][26];
int dx[4] = {-1,1,0,0}, dy[4] = {0,0,-1,1};
vector<string> board;

void dfs(int x, int y) {

    visited[x][y] = true;
    for(int i=0; i<4; i++) {
        int nx = x+dx[i];
        int ny = y+dy[i];
        if(nx < 0 || nx >= N || ny < 0 || ny >= N)  continue;
        if(board[nx][ny] == '1' && visited[nx][ny] == false) {
            cnt += 1;
            dfs(nx,ny);
        }
    }
}   

int main() {

    cin >> N;
    string str;
    for(int i=0; i<N; i++) {
        cin >> str;
        board.push_back(str);
    }
    memset(visited,false,sizeof(visited));
    vector<int> ans;
    for(int i=0; i<N; i++) {
        for(int j=0; j<N; j++) {
            if(board[i][j] == '1' && visited[i][j] == false) {
                cnt = 1;
                dfs(i,j);
                ans.push_back(cnt);
            }
        }
    }
    sort(ans.begin(),ans.end());
    cout << ans.size() << endl;
    for(int i=0; i<ans.size(); i++) {
        cout << ans[i] << endl;
    }

    return 0;
}